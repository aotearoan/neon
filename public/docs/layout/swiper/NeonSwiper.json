{
  "displayName": "NeonSwiper",
  "description": "Automatically handle horizontally overflowing content by placing it in a NeonSwiper component allowing for smooth\nhorizontal scrolling with all input devices.",
  "tags": {},
  "exportName": "default",
  "props": [
    {
      "name": "fade",
      "description": "Display fade in / fade out styling when there is an overflow.",
      "type": {
        "name": "boolean"
      },
      "defaultValue": {
        "func": false,
        "value": "true"
      }
    }
  ],
  "slots": [
    {
      "name": "default",
      "description": "contents of swiper with which to apply overflow scrolling."
    }
  ]
}