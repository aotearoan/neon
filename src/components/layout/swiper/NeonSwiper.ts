import { defineComponent, onMounted, onUnmounted, ref } from 'vue';

/**
 * Automatically handle horizontally overflowing content by placing it in a NeonSwiper component allowing for smooth
 * horizontal scrolling with all input devices.
 */
export default defineComponent({
  name: 'NeonSwiper',
  props: {
    /**
     * Display fade in / fade out styling when there is an overflow.
     */
    fade: { type: Boolean, default: true },
  },
  setup() {
    const scrollable = ref<HTMLElement | null>(null);
    const isOverflowing = ref<boolean>(false);
    const isScrollStart = ref<boolean>(false);
    const isScrollEnd = ref<boolean>(false);

    const handleScroll = () => {
      if (scrollable.value) {
        isScrollStart.value = scrollable.value.scrollLeft <= 0;
        isScrollEnd.value =
          scrollable.value.scrollLeft + scrollable.value.clientWidth > scrollable.value.scrollWidth - 1;
      } else {
        isScrollStart.value = false;
        isScrollEnd.value = false;
      }
    };

    const handleResize = () => {
      if (scrollable.value) {
        isOverflowing.value = scrollable.value.scrollWidth > scrollable.value.clientWidth;
      } else {
        isOverflowing.value = false;
      }

      handleScroll();
    };

    onMounted(() => {
      window.addEventListener('resize', handleResize, { passive: true });
      scrollable.value?.addEventListener('scroll', handleScroll, { passive: true });
      handleResize();
    });

    onUnmounted(() => {
      window.removeEventListener('resize', handleResize);
      scrollable.value?.removeEventListener('scroll', handleScroll);
    });

    return {
      isOverflowing,
      isScrollStart,
      isScrollEnd,
      scrollable,
    };
  },
});
