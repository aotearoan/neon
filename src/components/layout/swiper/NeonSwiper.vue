<template>
  <div
    :class="{
      'neon-swiper--fade': fade,
      'neon-swiper--overflowing': isOverflowing,
      'neon-swiper--start': isScrollStart,
      'neon-swiper--end': isScrollEnd,
    }"
    class="neon-swiper"
  >
    <div ref="scrollable" class="neon-swiper__container">
      <!-- @slot contents of swiper with which to apply overflow scrolling. -->
      <slot></slot>
    </div>
    <div class="neon-swiper__fade-out"></div>
  </div>
</template>

<script lang="ts" src="./NeonSwiper.ts"></script>
