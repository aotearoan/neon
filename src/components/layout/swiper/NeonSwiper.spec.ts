import type { RenderResult } from '@testing-library/vue';
import { render } from '@testing-library/vue';
import NeonSwiper from './NeonSwiper.vue';

describe('NeonSwiper', () => {
  let harness: RenderResult;

  beforeEach(() => {
    harness = render(NeonSwiper, {
      props: {},
      slots: { default: '<p>test</p>' },
    });
  });

  it('renders slot contents', () => {
    // given
    const { html } = harness;
    // when / then
    expect(html()).toMatchSnapshot();
  });

  it('applies fade style by default', () => {
    // given
    const { container } = harness;
    // when / then
    expect(container.querySelector('.neon-swiper--fade')).toBeDefined();
  });

  it('does not apply fade style when false', async () => {
    // given
    const { container, rerender } = harness;
    await rerender({ fade: false });
    // when / then
    expect(container.querySelector('.neon-swiper--fade')).toBeNull();
  });
});
