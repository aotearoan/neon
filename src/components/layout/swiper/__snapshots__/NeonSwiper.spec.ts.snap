// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeonSwiper renders slot contents 1`] = `
<div class="neon-swiper--fade neon-swiper--start neon-swiper--end neon-swiper">
  <div class="neon-swiper__container">
    <!-- @slot contents of swiper with which to apply overflow scrolling. -->
    <p>test</p>
  </div>
  <div class="neon-swiper__fade-out"></div>
</div>
`;
